### ISSUE TITLE
Connection pooling for Photon.js

### ISSUE THREAD

**Issue by nikolasburk** [prisma/photonjs#228]
Photon.js currently doesn't implement connection pooling at the moment (you can read more about managing connections in serverless environments in the [docs](https://github.com/prisma/prisma2/blob/master/docs/photon/deployment.md).

As of now, the recommended workaround is to use a tool like [PgBouncer](https://pgbouncer.github.io/faq.html). We are further exploring some options, such as [enabling a "DB proxy server"](https://github.com/prisma/prisma2/issues/370) (e.g. using a specific generator that generates a Docker image to host Photon's query engine) that manages the connection pool for you (similar to the the Prisma 1 architecture).

Also note that there some cloud offerings start to have solutions for connection pooling out-of-the-box, such as [AWS Aurora](https://aws.amazon.com/blogs/aws/new-data-api-for-amazon-aurora-serverless/).

**Comment by ScottAgirs** [comment]
Hey folks, here's my chime-in on this

## Problem:
On some occasions (currently only tested in development (local) environment) Heroku database gets flooded with connections that don't close. On a hobby plan the limit is 20 connections which can be max out in just few interactions with the app and results in a fatal error.

_The Error_ occurs (currently only this case is tested), when creating a new user or logging in.

## The Error

### Error querying the database: db error: FATAL: too many connections for role
![image](https://user-images.githubusercontent.com/11737572/68357280-cc64fa00-00e2-11ea-8cae-fca583ee3c2a.png)

The resolver `resolver.js`:
```javascript
export const LoginMutation = extendType({
  type: 'Mutation',
  definition(t) {
    t.field('login', {
      type: 'AuthPayload',
      args: {
        email: stringArg({ nullable: false }),
        password: stringArg({ nullable: false }),
      },
      resolve: async (_parent, { email, password }, ctx) => {
        const currentUser = await ctx.photon.users.findOne({
          where: {
            email,
          },
        })

        ..

        const token = "string"

        ctx.response.cookie('token', token, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 365,
        })

        return { currentUser }
      },
    })
  },
})
```

## Environment
- Prisma Framework (`prisma2@2.0.0-preview015, binary version: 20b6dc13949cccccfef5be07c0be7a3d7c858abe`) deployed to Zeit Now V2
- Postgres database deployed to Heroku
- Apollo Client 3.0 on the client

## Reproduction
Currently have not tried producing more than case in which I get this happens, but the one that shows consistent 

Also tried to just wait it out to see if it would close after a while, however, the connections remain active even after prolonged period of time (hours that is).

## Solutions?
Currently I need restart the server to fix.
Perhaps there is a function that can explicitly close the connection or something that I've missed?

Fingers crossed for solutions/easy workarounds to this 🤞

**Comment by mruzekw** [comment]
Is there documentation yet on how to connect through the AWS Aurora Serverless Data API? Thanks.

**Comment by brandonweiss** [comment]
Is it just me or is there some other bug with the connection pool (unrelated to serverless environments)? I seem to be hitting the same problem as @ScottAgirs. I’m using Heroku Postgres (Hobby, 20 connections) and running one instance of my application locally. Intermittently I’ll get a “too many connections” error and the only way to solve it is to kill the server and restart.

Is there some gotcha I missed? It feels like I shouldn’t be hitting the limit 🤔

**Comment by timsuchanek** [comment]
This is not really a bug, but we just open a bunch of connections in a connection pool.
You can limit them with appending `?connection_limit=1` to your connection string.
Here you can read more about it: https://www.prisma.io/docs/reference/database-connectors/postgresql#connection-url

**Comment by brandonweiss** [comment]
Huh 🤔 I’m on a MacBook Pro with 4 cores, so `num_physical_cpus * 2 + 1` would be 9, which is lower than the number of connections on the Heroku database, yet I’m still getting “too many connections” errors.

I was trying to solve this problem a few days ago and stumbled upon an issue that mentioned the `connection_limit` param in the connection string, so I tried that and set it to 10, but I still got errors…

It might be something I’m doing wrong on my end. The app is a Next.js app (a few days ago I ported it from Razzle to Next.js). The errors happened with both frameworks. In both cases I’m instantiating a new `PrismaClient` and then passing it into a graphQL server as “context” so I can make requests in resolvers using the same client. Both Razzle and Next.js have dev servers that detect when a change is made and reload all the app code so you don’t have to restart the server. Is there some gotcha I’m not seeing with my setup? And/or is it possible some state is being shared across instances in the Prisma client? 🤔

**Comment by Emetrop** [comment]
I have same problem. It creates with every query/mutation a new connection, setting up connection_limit solves nothing, it just ignores the number.

**Comment by nikolasburk** [comment]
Hey @brandonweiss, two questions:

> I’m instantiating a new PrismaClient and then passing it into a graphQL server as “context” so I can make requests in resolvers using the same client

Are you instantiating a new `PrismaClient` _per request_ or do you generally only have a single instance? (the latter is the recommended approach)

Also, are you calling `disconnect` in your code?

**Comment by Sytten** [comment]
@nikolasburk Not sure the recommended way is to call disconnect since that will kill the query engine process for each request. Generally creating a global `PrismaClient` with a `connection_limit` of 1 is sufficient, but you need to limit the number of parallel lambdas that can span (called reserved concurrency in AWS). Otherwise you have to use their new [proxy](https://aws.amazon.com/rds/proxy/) or pgBouncer. That is not a problem related to prisma as much as a general serverless problem.

**Comment by nikolasburk** [comment]
> @nikolasburk Not sure the recommended way is to call disconnect since that will kill the query engine process for each request. Generally creating a global `PrismaClient` with a `connection_limit` of 1 is sufficient, but you need to limit the number of parallel lambdas that can span (called reserved concurrency in AWS). Otherwise, you have to use their new [proxy](https://aws.amazon.com/rds/proxy/) or pgBouncer. That is not a problem related to Prisma as much as a general serverless problem.

That all seems correct to me for serverless environments – but I believe in this case we're talking about a long-running Node.js process that's running locally on the user's machine so I believe calling `disconnect()` when the application terminates is a good practice. Furthermore, it's indeed recommended to only create a single `PrisamClient` instance rather than creating one for each request. Basically this:

**Yes**

```
const prisma = new PrismaClient()
const server = new GraphQLServer({
  schema,
  context: {
    prisma
  },
})
```

**No**

```ts
const server = new GraphQLServer({
  schema,
  context: {
    prisma: new PrismaClient()
  },
})
```

**Comment by brandonweiss** [comment]
@nikolasburk I’m instantiating it once and passing it into the graphQL server context similar to your “yes” example. And I’m not calling disconnect anywhere. I’ll post my exact code:

```js
const prisma = new PrismaClient()

let apolloServer = new ApolloServer({
  context: async ({ req: { session } }) => {
    let userID = session.get("userID")

    return {
      userID,
      prisma,
    }
  },
})
```

Context is taking a function in order to have access to the request/session, but I’m not instantiating a new client per request—it should always be the same instance, right?

I assume the reason killing/restarting the server works is because Prisma then will call disconnect and flush the connections?

I could be completely misunderstanding how the client and connections works, but if instantiating a new client per-request is bad (because the connections hang around?) … then wouldn’t a dev server reloading essentially recreate that effect? When Next.js trashes the require cache and re-requires the graphQL server, it will create a new client. I assume over time connections must timeout and die, but if someone were to make a lot of changes/reloads really quickly, presumably this would use up all the connections?

If that’s the case, I would sort of expect this to be a bigger problem, but perhaps most people are developing against a local (or remote) database with a much higher connection limit?

**Comment by Sytten** [comment]
@brandonweiss Yes if not for the lifecycle handling done inside the client itself. The recommended way is to call disconnect to be sure, but the client catches some OS signals and kills the child process. See https://github.com/prisma/prisma/blob/3cb0854e771f4c6daf9be007bc199e5c1405dd67/src/packages/engine-core/src/NodeEngine.ts#L790

**Comment by nikolasburk** [comment]
> I assume the reason killing/restarting the server works is because Prisma then will call disconnect and flush the connections?
> then wouldn’t a dev server reloading essentially recreate that effect?

We do want to work on [automatic `disconnect` behavior](https://github.com/prisma/prisma-client-js/issues/540) (and most of the time automatic `disconnect` already should work today) but for now you probably want to be on the safe side and call `disconnect` explicitly in your code so that it gets. called before your dev server restarts.

**Comment by Emetrop** [comment]
I think that my and @brandonweiss problem with heroku should solve adding pgbouncer buildpack 

https://devcenter.heroku.com/articles/on-dyno-postgres-connection-pooling

**Comment by Emetrop** [comment]
So I was right. With heroku is needed to do this:

1. add pgbouncer buldpack - `heroku buildpacks:add heroku/pgbouncer`
2. create `Procfile` file (without extension and with uppercase P) in root of your project
3. add config to `Procfile` - for me with node.js project it's this one line `web: bin/start-pgbouncer npm run start`

**Comment by brandonweiss** [comment]
I was playing around with my setup and discovered that Next.js only rebuilds _certain_ files in development mode. The file I was instantiating the client in was one of those that gets rebuilt on every change refresh.

I moved the client instantiation into a file in a directory that Next.js doesn’t watch.

```js
// lib/prisma.js

const { PrismaClient } = require("@prisma/client")

export default new PrismaClient()
```

It worked fine for a while, but then I got another “too many connections“ error. I thought about it for a bit and then I remembered someone saying that the client can potentially open lots of connections and it can be constrainted by using the `connection_limit` query param on the database connection string.

I set mine to 20, as that’s the maximum number of connections for a Heroku dev instance and I was feeling like I’d solved it, and then I got hit with another “too many connections” error.

At this point I have no idea what’s left to try to debug this. I’ve got a log statement right before/after the prisma client gets instantiated and it only ever happens once, right as the dev server starts up, so I’m like 99.9% sure I don’t have multiple clients.

Is it possible there’s a bug in the connection pooling behavior? Any suggestions on what else I might try?

I can add PgBouncer and that will probably fix it, but I’d much rather solve the root problem if at all possible. Thanks! 🙏

**Comment by Sytten** [comment]
Can you verify processes named `query-engine` and see if you have multiple spawned? That would be the real indicator if clients are spawned or not. Also 20 connections is also a lot for a single client, try setting that to 2-3 it should be sufficient for dev.

**Comment by brandonweiss** [comment]
Interesting… I just looked and I had 6. I restarted the dev server and I immediately had 2. Made some changes and refreshed a bit and now I have 4.

Ugh, OK, I just tested it again and the file where I instantiated the Prisma client that I _thought_ wasn’t being reloaded _is_ being reloaded 😭 

I found an [issue on the Next.js repo](https://github.com/zeit/next.js/issues/7811) where people were having similar problems. So it seems like this is going to be an issue with Prisma and pretty much any dev server. A dev server probably needs to have the ability to ignore reloading certain files, which is slightly different from the ability to ignore _watching_ certain files. The latter is common, but the former is something I’m not sure I can remember seeing before.

From a developer experience perspective, this is a tough one. When the docs say not to instantiate a new client for every request, most people will probably understand what the right thing to do is, especially if there is a code example explaining it. But guarding against a problem like this is going to be pretty hard. Neither Prisma nor Next.js are doing anything wrong, but in combination there’s a really unexpected result.

It might be worth adding a note to the documentation about this particular scenario. Since some (or many) dev servers might have no way of not reloading a certain file, it might be worth including this snippet as an example of how to work around it.

```js
import { PrismaClient } from "@prisma/client"

let prisma

if (process.env.NODE_ENV === "production") {
  prisma = new PrismaClient()
} else {
  if (!global.prisma) {
    global.prisma = new PrismaClient()
  }

  prisma = global.prisma
}

export default prisma
```

Additionally, it would have been really awesome to get a proactive warning that I was instantiating multiple Prisma clients. You could maybe do something clever and keep track of instantiated instances internally and then log a warning if someone creates more than one? Most people probably don’t intend to create multiple clients so a warning could be really helpful. You could also provide a way for people to ignore the warning if they intentionally mean to instantiate multiple clients for some reason? Just some thoughts!

**Comment by Sytten** [comment]
@brandonweiss Agreed, feel free to open a feature request on the multiple clients thing. I would say it probably is not a big priority because eventually prisma will move to a direct binding to rust so it shouldn't be as much as an issue as it is currently. 
@nikolasburk @pantharshit00 do you want another issue for the documentation too?

**Comment by pantharshit00** [comment]
Yes please, docs issue won't hurt and we can track that separately :)

**Comment by jeffdc** [comment]
I have been struggling with this as well. It is also a huge pain when running unit tests. Anyhow, I tried the advice given here (using global) and I am still getting hundreds of dangling connections (and related query engine processes). I have a single instance of Prisma that I use across my app (at least that I what I am trying to do!). In that file I implemented the code as mentioned in this thread but it is not helping. I am seeing console messages (`Development: Created DB connection.` - see code below) that shows that it is creating a new connection over and over again (after every change/compilation - Next.js). 

I also have a variation of this sad state of affairs and have to run `killall query-engine-darwin` with regularity:
```bash
jeff gallformers (descriptions●●) % ps -A | grep -c query-engine-darwin
258
```

Here is my connection code:
```TypeScript
/* eslint-disable @typescript-eslint/ban-ts-comment */
import { PrismaClient } from '@prisma/client';

let db: PrismaClient;
if (process.env.NODE_ENV === 'production') {
    db = new PrismaClient();
    console.log('Production: Created DB connection.');
} else {
    // @ts-ignore
    if (!global.db) {
        // @ts-ignore
        global.db = new PrismaClient();
        console.log('Development: Created DB connection.');
    }

    // @ts-ignore
    db = global.db;
}

db.$executeRaw('PRAGMA foreign_keys = ON');

export default db;
```

I had to resort to `--forceExit` on my Jest tests otherwise they would hang forever with an open DB connection. I can close it in `afterAll` but it only works if I wait at least 1 second in the Promise after each test suite for it to actually close. This is untenable as it makes the tests run very slowly. I am not sure if this is a Prisma or SQlite thing. That code looks like this:
```TypeScript
afterAll(async (done) => {
    await new Promise((resolve) => {
        setTimeout(() => {
            db.$disconnect();
            resolve();
        }, 1000);
    });
    done();
});
```

Anyhow, this is a pain in dev but hopefully will not bite me once I move this to production.

**Comment by Sytten** [comment]
@jeffdc For tests we use `--runInBand` so they dont run in parallel, apart from that we don't do anything particular since I don't think jest reloads modules once they are loaded once.
We don't even do `await prisma.$disconnect();`, but I didn't check if they leaked connections. Tests are really slow though, but thats fine for me since those are integration tests that are always slow.
